p4-16/psa/PSA.mdk

```diff
@@ -462,26 +464,41 @@ Example usage:
 
 ```
 parser P() {
-  Hash<crc16, bit<16>> h();
-  bit<16> hash_value = h.getHash(16, buffer, 100);
+  Hash<bit<16>>(HashAlgorithm.crc16) h;
+  bit<16> hash_value = h.getHash(buffer);
 }
 ```
 
 Parameters:
 
-- Algo The algorithm to use for computation (@see HashAlgorithm).
+- algo The algorithm to use for computation (@see HashAlgorithm).
 - O    The type of the return value of the hash.
 
 ```
-extern Hash<Algo, O> {
+extern Hash<O> {
   /// Constructor
-  Hash();
-
-  /// compute the hash for data
+  Hash(HashAlgorithm algo);
+
+  /// Compute the hash for data.
+  /// @param data The data over which to calculate the hash.
+  /// @return The hash value.
+  O getHash<D>(in D data);
+
+  /// Compute the hash for data, with modulo by max, then add base.
+  /// @param base Minimum return value.
+  /// @param data The data over which to calculate the hash.
+  /// @param max The hash value is divided by max to get modulo.
+  ///        An implementation may limit the largest value supported,
+  ///        e.g. to a value like 32, or 256.
+  /// @return (base + (h % max)) where h is the hash value.
   O getHash<T, D, M>(in T base, in D data, in M max);
 }
 ```
 
+TBD: Should there be a `const` defined that specifies the maximum
```

Comment:
I think that maximum should be a target specific value. Of course there is a max, 2^sizeof(O)-1 ...
But, do we need separate types for base and max (i.e., T and M)? They are likely to be the same size ...